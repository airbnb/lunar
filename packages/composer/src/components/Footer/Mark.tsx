import React from 'react';
import useStyles from '@airbnb/lunar/lib/hooks/useStyles';
import { footerMarkStyleSheet } from '../../styles';

export type MarkProps = {
  children?: React.ReactNode;
};

export default function Mark({ children }: MarkProps) {
  const [styles, cx] = useStyles(footerMarkStyleSheet);

  return <mark className={cx(styles.mark)}>{children}</mark>;
}
