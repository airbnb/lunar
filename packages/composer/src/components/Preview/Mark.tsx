import React, { useRef } from 'react';
import useStyles from '@airbnb/lunar/lib/hooks/useStyles';
import { ProofreadConfig } from '../../types';
import { previewMarkStyleSheet } from '../../styles';

export type MarkProps = {
  children: string;
  error: ProofreadConfig;
  highlighted?: boolean;
  secondary?: boolean;
  selected?: boolean;
  onClick: (error: ProofreadConfig, top: number, left: number) => void;
};

export default function Mark({
  children,
  error,
  highlighted,
  secondary,
  selected,
  onClick,
}: MarkProps) {
  const [styles, cx] = useStyles(previewMarkStyleSheet);
  const ref = useRef<HTMLSpanElement | null>(null);
  const active = highlighted || selected;

  return (
    <button
      className={cx(styles.button)}
      type="button"
      onClick={() => {
        if (ref.current) {
          onClick(error, ref.current.offsetTop + ref.current.offsetHeight, ref.current.offsetLeft);
        }
      }}
    >
      <mark
        ref={ref}
        className={cx(
          styles.mark,
          active && styles.mark_highlight,
          secondary && styles.markSecondary,
          secondary && active && styles.markSecondary_highlight,
        )}
      >
        {children}
      </mark>
    </button>
  );
}
