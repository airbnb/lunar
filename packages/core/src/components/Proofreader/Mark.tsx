import React, { useRef, useEffect, useCallback } from 'react';
import useStyles from '../../hooks/useStyles';
import { ProofreadRuleMatch } from './types';
import { markStyleSheet } from './styles';

export type MarkProps = {
  children: string;
  error: ProofreadRuleMatch;
  highlighted?: boolean;
  secondary?: boolean;
  selected?: boolean;
  shadow?: boolean;
  onSelect: (error: ProofreadRuleMatch, top: number, left: number) => void;
};

export default function Mark({
  children,
  error,
  highlighted,
  secondary,
  selected,
  shadow = false,
  onSelect,
}: MarkProps) {
  const [styles, cx] = useStyles(markStyleSheet);
  const ref = useRef<HTMLSpanElement | null>(null);
  const active = highlighted || selected;

  const handleSelect = useCallback(() => {
    if (ref.current) {
      onSelect(error, ref.current.offsetTop + ref.current.offsetHeight, ref.current.offsetLeft);
    }
  }, [error, onSelect]);

  // When being used as a shadow, the mark cannot be clicked.
  // So we automatically fire the handler when selected state changes.
  useEffect(() => {
    if (shadow && selected) {
      handleSelect();
    }
  }, [shadow, selected, handleSelect]);

  return (
    <button className={cx(styles.button)} type="button" onClick={handleSelect}>
      <mark
        ref={ref}
        className={cx(
          styles.mark,
          active && styles.mark_highlight,
          secondary && styles.markSecondary,
          secondary && active && styles.markSecondary_highlight,
        )}
      >
        {children}
      </mark>
    </button>
  );
}
