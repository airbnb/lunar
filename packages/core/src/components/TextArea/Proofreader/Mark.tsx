import React from 'react';
import withStyles, { WithStylesProps } from '../../../composers/withStyles';

export type Props = {
  children: NonNullable<React.ReactNode>;
  selected: boolean;
  onSelect: (top: number, left: number) => void;
};

class Mark extends React.PureComponent<Props & WithStylesProps> {
  ref = React.createRef<HTMLSpanElement>();

  componentDidMount() {
    this.handleSelect();
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.selected !== prevProps.selected) {
      this.handleSelect();
    }
  }

  private handleSelect = () => {
    const { ref } = this;

    if (this.props.selected && ref.current) {
      this.props.onSelect(ref.current.offsetTop + ref.current.offsetHeight, ref.current.offsetLeft);
    }
  };

  render() {
    const { cx, children, selected, styles } = this.props;

    return (
      <mark className={cx(styles.mark, selected && styles.mark_selected)} ref={this.ref}>
        {children}
      </mark>
    );
  }
}

export default withStyles(({ color, ui }) => ({
  mark: {
    position: 'relative',
    color: 'transparent',
    backgroundColor: 'transparent',
    opacity: 0.75,
    padding: 1,
    margin: -1,
    transition: 'opacity .2s, background .2s',

    '::after': {
      position: 'absolute',
      display: 'block',
      content: '""',
      bottom: 0,
      left: 0,
      right: 0,
      height: 2,
      borderRadius: ui.borderRadius,
      backgroundColor: color.core.danger[2],
    },
  },

  mark_selected: {
    opacity: ui.disabledOpacity,
    borderTopLeftRadius: ui.borderRadius,
    borderTopRightRadius: ui.borderRadius,
    backgroundColor: color.core.danger[2],
  },
}))(Mark);
