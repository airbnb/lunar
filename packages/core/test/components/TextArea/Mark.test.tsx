import React from 'react';
import { mountWithStyles, shallowWithStyles } from '@airbnb/lunar-test-utils';
import Mark from '../../../src/components/TextArea/Proofreader/Mark';

describe('<Mark />', () => {
  it('renders', () => {
    const wrapper = shallowWithStyles(
      <Mark selected={false} onSelect={() => {}}>
        Word
      </Mark>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders selected', () => {
    const wrapper = shallowWithStyles(
      <Mark selected onSelect={() => {}}>
        Word
      </Mark>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('triggers `onSelect` on mount if selected', () => {
    const spy = jest.fn();

    mountWithStyles(
      <Mark selected onSelect={spy}>
        Word
      </Mark>,
    );

    expect(spy).toHaveBeenCalled();
  });

  it('doesnt trigger `onSelect` on mount if not selected', () => {
    const spy = jest.fn();

    mountWithStyles(
      <Mark selected={false} onSelect={spy}>
        Word
      </Mark>,
    );

    expect(spy).not.toHaveBeenCalled();
  });

  it('triggers `onSelect` if updated to selected', () => {
    const spy = jest.fn();
    const wrapper = mountWithStyles(
      <Mark selected={false} onSelect={spy}>
        Word
      </Mark>,
    );

    expect(spy).not.toHaveBeenCalled();

    wrapper.setProps({
      selected: true,
    });

    expect(spy).toHaveBeenCalled();
  });

  it('doesnt trigger `onSelect` if updated from selected', () => {
    const spy = jest.fn();
    const wrapper = mountWithStyles(
      <Mark selected onSelect={spy}>
        Word
      </Mark>,
    );

    expect(spy).toHaveBeenCalledTimes(1);

    wrapper.setProps({
      selected: false,
    });

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
