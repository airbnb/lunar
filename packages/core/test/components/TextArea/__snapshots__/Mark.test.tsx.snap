// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Mark /> renders 1`] = `
<mark
  className="mark_jc2c4c"
>
  Word
</mark>
`;

exports[`<Mark /> renders selected 1`] = `
<mark
  className="mark_jc2c4c-o_O-mark_selected_1arrkjl"
>
  Word
</mark>
`;
